lightning/src/util/test_utils.rs

```diff
@@ -1731,6 +1736,132 @@ impl Logger for TestLogger {
 	}
 }
 
+pub struct TestTracerLayer;
+
+#[derive(Debug, Default)]
+struct NodeInfo {
+	node_id: Option<String>,
+}
+
+struct NodeVisitor {
+	node_id: Option<String>,
+}
+
+impl NodeVisitor {
+	fn new() -> Self {
+		Self { node_id: None }
+	}
+}
+
+impl Visit for NodeVisitor {
+	fn record_str(&mut self, field: &Field, value: &str) {
+		if field.name() == "node_id" {
+			self.node_id = Some(value.to_string());
+		}
+	}
+	fn record_debug(&mut self, field: &Field, value: &dyn fmt::Debug) {
+		// if field.name() == "node_id" {
+		// 	self.node_id = Some(format!("{:?}", value));
+		// }
+	}
+}
+
+struct MessageVisitor {
+	message: Option<String>,
+}
+
+impl MessageVisitor {
+	fn new() -> Self {
+		Self { message: None }
+	}
+}
+
+impl Visit for MessageVisitor {
+	fn record_debug(&mut self, field: &Field, value: &dyn fmt::Debug) {
+		if field.name() == "message" {
+			self.message = Some(format!("{:?}", value));
+		}
+	}
+}
+
+impl<S> Layer<S> for TestTracerLayer
+where
+	S: Subscriber + for<'lookup> LookupSpan<'lookup>,
+{
+	fn on_new_span(
+		&self, attrs: &tracing::span::Attributes<'_>, id: &tracing::span::Id,
+		ctx: tracing_subscriber::layer::Context<'_, S>,
+	) {
+		let mut visitor = NodeVisitor::new();
+		attrs.record(&mut visitor);
+
+		let span = ctx.span(id).expect("span exists");
+		let mut extensions = span.extensions_mut();
+		let mut info = NodeInfo::default();
+		info.node_id = visitor.node_id;
+		extensions.insert(info);
+	}
+
+	fn on_record(
+		&self, id: &tracing::span::Id, values: &tracing::span::Record<'_>,
+		ctx: tracing_subscriber::layer::Context<'_, S>,
+	) {
+		let span = ctx.span(id).expect("span exists");
+		let mut visitor = NodeVisitor::new();
+		values.record(&mut visitor);
+
+		if let Some(node_id) = visitor.node_id {
+			let mut extensions = span.extensions_mut();
+			if let Some(info) = extensions.get_mut::<NodeInfo>() {
+				info.node_id = Some(node_id);
+			}
+		}
+	}
+
+	fn on_event(&self, event: &Event<'_>, ctx: tracing_subscriber::layer::Context<'_, S>) {
+		let meta = event.metadata();
+		let level = *meta.level();
+
+		let mut visitor = MessageVisitor::new();
+		event.record(&mut visitor);
+		let message = visitor.message.unwrap_or_else(|| "<no message>".to_string());
+
+		// Walk up the span tree to find a node_id
+		let mut node_id = "?".to_string();
+		if let Some(span) = ctx.event_span(event) {
+			let mut current = Some(span);
+			while let Some(s) = current {
+				let extensions = s.extensions();
+				if let Some(info) = extensions.get::<NodeInfo>() {
```

Comment:
Don't ask too many questions about how this code is getting the node id.... Testing only...